import() in d.ts files can't be resolved if used as template argument of other import()'ed type

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  2.9.1

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** declaration

**Code**
test1.d.ts
```ts
export interface T<P> {
    a: P;
}
```
test2.d.ts
```ts
export declare const theme: { a: string }
```

test3.d.ts
```ts
export declare const a: import("./test1").T<import("./test2").theme>;
```

**Expected behavior:**
Compiles correctly

**Actual behavior:**
```
test3.d.ts:1:45 - error TS2307: Cannot find module './test2'.
```

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> 

**Related Issues:** <!-- Did you find other bugs that looked similar? -->

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -1819,11 +1819,9 @@ namespace ts {
                 else if (isLiteralImportTypeNode(node)) {
                     imports = append(imports, node.argument.literal);
                 }
-                else {
-                    collectDynamicImportOrRequireCallsForEachChild(node);
-                    if (hasJSDocNodes(node)) {
-                        forEach(node.jsDoc, collectDynamicImportOrRequireCallsForEachChild);
-                    }
+                collectDynamicImportOrRequireCallsForEachChild(node);
+                if (hasJSDocNodes(node)) {
+                    forEach(node.jsDoc, collectDynamicImportOrRequireCallsForEachChild);
                 }
             }
 

--- a/tests/baselines/reference/importUsedInGenericImportResolves.js
+++ b/tests/baselines/reference/importUsedInGenericImportResolves.js
@@ -1,0 +1,17 @@
+//// [tests/cases/compiler/importUsedInGenericImportResolves.ts] ////
+
+//// [test1.d.ts]
+export interface T<P> {
+    a: P;
+}
+
+//// [test2.d.ts]
+export declare const theme: { a: string }
+
+//// [test3.ts]
+export const a: import("./test1").T<typeof import("./test2").theme> = null as any;
+
+//// [test3.js]
+"use strict";
+exports.__esModule = true;
+exports.a = null;

--- a/tests/baselines/reference/importUsedInGenericImportResolves.symbols
+++ b/tests/baselines/reference/importUsedInGenericImportResolves.symbols
@@ -1,0 +1,21 @@
+=== tests/cases/compiler/test1.d.ts ===
+export interface T<P> {
+>T : Symbol(T, Decl(test1.d.ts, 0, 0))
+>P : Symbol(P, Decl(test1.d.ts, 0, 19))
+
+    a: P;
+>a : Symbol(T.a, Decl(test1.d.ts, 0, 23))
+>P : Symbol(P, Decl(test1.d.ts, 0, 19))
+}
+
+=== tests/cases/compiler/test2.d.ts ===
+export declare const theme: { a: string }
+>theme : Symbol(theme, Decl(test2.d.ts, 0, 20))
+>a : Symbol(a, Decl(test2.d.ts, 0, 29))
+
+=== tests/cases/compiler/test3.ts ===
+export const a: import("./test1").T<typeof import("./test2").theme> = null as any;
+>a : Symbol(a, Decl(test3.ts, 0, 12))
+>T : Symbol(T, Decl(test1.d.ts, 0, 0))
+>theme : Symbol(theme, Decl(test2.d.ts, 0, 20))
+

--- a/tests/baselines/reference/importUsedInGenericImportResolves.types
+++ b/tests/baselines/reference/importUsedInGenericImportResolves.types
@@ -1,0 +1,23 @@
+=== tests/cases/compiler/test1.d.ts ===
+export interface T<P> {
+>T : T<P>
+>P : P
+
+    a: P;
+>a : P
+>P : P
+}
+
+=== tests/cases/compiler/test2.d.ts ===
+export declare const theme: { a: string }
+>theme : { a: string; }
+>a : string
+
+=== tests/cases/compiler/test3.ts ===
+export const a: import("./test1").T<typeof import("./test2").theme> = null as any;
+>a : import("tests/cases/compiler/test1").T<{ a: string; }>
+>T : import("tests/cases/compiler/test1").T<P>
+>theme : any
+>null as any : any
+>null : null
+

--- a/tests/cases/compiler/importUsedInGenericImportResolves.ts
+++ b/tests/cases/compiler/importUsedInGenericImportResolves.ts
@@ -1,0 +1,10 @@
+// @filename: test1.d.ts
+export interface T<P> {
+    a: P;
+}
+
+// @filename: test2.d.ts
+export declare const theme: { a: string }
+
+// @filename: test3.ts
+export const a: import("./test1").T<typeof import("./test2").theme> = null as any;

